lib/document.js

```diff
@@ -4582,6 +4582,27 @@ Document.prototype.getChanges = function() {
   return changes;
 };
 
+/**
+ * Returns a copy of this document with a deep clone of `_doc` and `$__`.
+ *
+ * @return {Document} a copy of this document
+ * @api private
+ * @method $clone
+ * @memberOf Document
+ * @instance
+ */
+
+Document.prototype.$clone = function() {
+  const clonedDoc = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
+  if (this._doc) {
+    clonedDoc._doc = clone(this._doc);
+  }
+  if (this.$__) {
+    clonedDoc.$__ = Object.assign(Object.create(Object.getPrototypeOf(this.$__)), this.$__);
```

Comment:
We would need to do a deep clone on `$__`, that may contain deeply nested properties.